controllers/openstackfloatingippool_controller.go

```diff
@@ -0,0 +1,471 @@
+	ipAddresses := &ipamv1.IPAddressList{}
+	if err := r.Client.List(ctx, ipAddresses, client.InNamespace(pool.Namespace), client.MatchingFields{infrav1.OpenStackFloatingIPPoolNameIndex: pool.Name}); err != nil {
+		return err
+	}
+
+	networkingService, err := networking.NewService(scope)
+	if err != nil {
+		return err
+	}
+	pool.Status.ClaimedIPs = []string{}
+	if pool.Status.AvailableIPs == nil {
+		pool.Status.AvailableIPs = []string{}
+	}
+
+	for i := 0; i < len(ipAddresses.Items); i++ {
+		ipAddress := &(ipAddresses.Items[i])
+		if ipAddress.ObjectMeta.DeletionTimestamp.IsZero() {
+			pool.Status.ClaimedIPs = append(pool.Status.ClaimedIPs, ipAddress.Spec.Address)
+			continue
+		}
+
+		if controllerutil.ContainsFinalizer(ipAddress, infrav1.DeleteFloatingIPFinalizer) {
+			if pool.Spec.ReclaimPolicy == infrav1.ReclaimDelete && !contains(pool.Spec.PreAllocatedFloatingIPs, ipAddress.Spec.Address) {
+				if err = networkingService.DeleteFloatingIP(pool, ipAddress.Spec.Address); err != nil {
+					return fmt.Errorf("delete floating IP %q: %w", ipAddress.Spec.Address, err)
+				}
+			} else {
+				pool.Status.AvailableIPs = append(pool.Status.AvailableIPs, ipAddress.Spec.Address)
+			}
+		}
+		controllerutil.RemoveFinalizer(ipAddress, infrav1.DeleteFloatingIPFinalizer)
+		if err := r.Client.Update(ctx, ipAddress); err != nil {
+			return err
+		}
+	}
+	unclaimedPreAllocatedIPs := diff(pool.Spec.PreAllocatedFloatingIPs, pool.Status.ClaimedIPs)
+	unclaimedIPs := union(pool.Status.AvailableIPs, unclaimedPreAllocatedIPs)
+	pool.Status.AvailableIPs = diff(unclaimedIPs, pool.Status.FailedIPs)
+	return nil
+}
+
+func (r *OpenStackFloatingIPPoolReconciler) getIP(scope scope.Scope, pool *infrav1.OpenStackFloatingIPPool) (string, error) {
+	// There's a potential leak of IPs here, if the reconcile loop fails after we claim an IP but before we create the IPAddress object.
+	var ip string
+
+	networkingService, err := networking.NewService(scope)
+	if err != nil {
+		scope.Logger().Error(err, "Failed to create networking service")
+		return "", err
+	}
+
+	// Get tagged floating IPs and add them to the available IPs if they are not present in either the available IPs or the claimed IPs
+	// This is done to prevent leaking floating IPs if to prevent leaking floating IPs if the floating IP was created but the IPAddress object was not
+	taggedIPs, err := networkingService.GetFloatingIPsByTag(pool.GetFloatingIPTag())
+	if err != nil {
+		scope.Logger().Error(err, "Failed to get floating IPs by tag", "pool", pool.Name)
+		return "", err
+	}
+	for _, taggedIP := range taggedIPs {
+		if contains(pool.Status.AvailableIPs, taggedIP.FloatingIP) || contains(pool.Status.ClaimedIPs, taggedIP.FloatingIP) {
+			continue
+		}
+		scope.Logger().Info("Tagged floating IP found that was not known to the pool, adding it to the pool", "ip", taggedIP.FloatingIP)
+		pool.Status.AvailableIPs = append(pool.Status.AvailableIPs, taggedIP.FloatingIP)
+	}
```

Comment:
We probably don't want to do this unconditionally in the IPAddressClaim loop for every unallocated IPAddressClaim. Suggest we could do it only if len(AvailableIPs) == 0?